### PR TITLE
Add objc auth demo

### PR DESCRIPTION
**Pull request by jcanizales**
I extracted a few TODOs from writing this code:
* ~~Put the response headers in the `NSError`.~~
* ~~Let the common user ignore the case of duplicate header names~~. Louis suggests to ~~have `responseMetadata` drop duplicates, and~~ add an array (`responseMetadataArray`?) of all received metadata elements (`GRPCMetadataEntry`?).
* Make calls retriable. A very natural way is to let people copy a `GRXWriter` before it's started. Once we have that, ease the case of handling expired/revoked token: https://github.com/grpc/grpc/issues/2091
* ~~Let assign an immutable dictionary to `requestMetadata`.~~
* ~~Initialize `requestMetadata` to an empty `NSMutableDictionary` (instead of `nil`), so one can just go and set values to its keys.~~
* Include an OAuth2 category of `GRPCCall` to let people just write `call.accessToken = token`: https://github.com/grpc/grpc/issues/2089

### CODE REVIEW

**Comment by murgatroid99** [comment]
Code LGTM

**Comment by jcanizales** [comment]
@LisaFC , can you review the README and the addition to the common auth doc?

**Comment by jcanizales** [comment]
Ping.

**Comment by murgatroid99** [comment]
I'm not sure what happened, but it looks like there was a merge conflict.

**Comment by jcanizales** [comment]
Done. Apparently Git's merge doesn't like it when two different things are appended at the end of a file (which should go last?).

**Comment by jcanizales** [comment]
@LisaFC can you review the documentation?

**Comment by jcanizales** [comment]
Ping. We're blocked on this sample to make a public announcement about Objective-C.

**Comment by LisaFC** [comment]
I'll take a look!

File: objective-c/auth_sample/README.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+- Initialize and configure a remote call object before the RPC is started.
+- Set request metadata elements on a call, which are semantically equivalent to HTTP request
+headers.
+- Read response metadata from a call, which is equivalent to HTTP response headers and trailers.
+
+It assumes you know the basics on how to make gRPC API calls using the Objective-C client library,
+as shown in the [Hello World](https://github.com/grpc/grpc-common/tree/master/objective-c/helloworld)
+or [Route Guide](https://github.com/grpc/grpc-common/tree/master/objective-c/route_guide) tutorials,
+and familiarity with OAuth2 concepts like _access token_.
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: and familiarity with -> and are familiar with

File: objective-c/auth_sample/README.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+```shell
+$ git clone https://github.com/grpc/grpc-common.git
+```
+
+Then change your current directory to `grpc-common/objective-c/auth_sample`:
+```shell
+$ cd grpc-common/objective-c/auth_sample
+```
+
+Our example is a simple application with two views. The first one lets a user sign in and out using
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: Is "view" a common term that an Objective-C user would understand without further explanation?

File: objective-c/auth_sample/README.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+(Google's library is used in this example because the test gRPC service we are going to call expects
+Google account credentials, but neither gRPC nor the Objective-C client library is tied to any
+specific OAuth2 provider). The second view makes a gRPC request to the test server, using the
+access token obtained by the first view.
+
+Note: OAuth2 libraries need the application to register and obtain an ID from the identity provider
+(in the case of this example app, Google). The app's XCode project is configured using that ID, so
+you shouldn't copy this project as is for your own app: It would result in your app being identified
+in the consent screen as "gRPC-AuthSample", and not having access to real Google services. Instead,
+configure your XCode project following the [instructions here](https://developers.google.com/identity/sign-in/ios/).
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: configure your XCode project -> configure your own XCode project

File: objective-c/auth_sample/README.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Our example is a simple application with two views. The first one lets a user sign in and out using
+the OAuth2 flow of Google's [iOS SignIn library](https://developers.google.com/identity/sign-in/ios/).
+(Google's library is used in this example because the test gRPC service we are going to call expects
+Google account credentials, but neither gRPC nor the Objective-C client library is tied to any
+specific OAuth2 provider). The second view makes a gRPC request to the test server, using the
+access token obtained by the first view.
+
+Note: OAuth2 libraries need the application to register and obtain an ID from the identity provider
+(in the case of this example app, Google). The app's XCode project is configured using that ID, so
+you shouldn't copy this project as is for your own app: It would result in your app being identified
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: as is -> "as is"
for your own app: It -> for your own app: it

File: objective-c/auth_sample/README.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+The first view, `SelectUserViewController.h/m`, asks you to sign in with your Google account, and to
+give the "gRPC-AuthSample" app the following permissions:
+
+- View your email address.
+- View your basic profile info.
+- "Test scope for access to the Zoo service".
+
+This last permission, corresponding to the scope `https://www.googleapis.com/auth/xapi.zoo` doesn't
+grant any real capability: It's only used for testing. You can log out at any moment.
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: capability: It's -> capability: it's
at any moment -> at any time

File: objective-c/auth_sample/README.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+```protobuf
+option objc_class_prefix = "AUTH";
+
+service TestService {
+  rpc UnaryCall(Request) returns (Response);
+}
+```
+
+A `unaryCallWithRequest:handler:` method is generated, with which you're already familiar, is
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: duplicated "is generated"

File: objective-c/auth_sample/README.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+generated for the `AUTHTestService` class:
+
+```objective-c
+[client unaryCallWithRequest:request handler:^(AUTHResponse *response, NSError *error) {
+  ...
+}];
+```
+
+And a second `RPCToUnaryCallWithRequest:handler:` method is generated, which returns a
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: And a second `RPCToUnaryCallWithRequest:handler:` method is generated ->
In addition, an `RPCToUnaryCallWithRequest:handler:` method is generated

File: objective-c/auth_sample/README.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+## Set request metadata of a call: Authorization header with an access token
+
+The `ProtoRPC` class has a `requestMetadata` property defined like this:
+
+```objective-c
+@property(nonatomic, readwrite) NSMutableDictionary *requestMetadata;
+```
+
+Setting it to a dictionary of metadata keys and values will have them sent on the wire when the call
+is started.
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: Should we add in a bit more about what metadata is, ie that it's info sent by the client to the server (and vice versa) about the call that's essentially opaque to gRPC itself? Though there'll eventually be info about it in the gRPC concepts doc that I haven't finished yet...

File: objective-c/auth_sample/README.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+If you have an access token, OAuth2 specifies it is to be sent in this format:
+
+```objective-c
+@{@"Authorization": [@"Bearer " stringByAppendingString:accessToken]}
+```
+
+<a name="response-metadata"></a>
+## Get response metadata of a call: Auth challenge header
+
+Analogously to the request metadata, the `ProtoRPC` class has a `responseMetadata` property defined
```
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: "Analogously" reads a bit oddly. How about 

The `ProtoRPC` class also has a `responseMetadata` property, analogous to the request metadata we just looked at. It's defined like this:

**Comment by LisaFC** [comment]
Tutorial LGTM, it's nice and clear.

File: objective-c/auth_sample/README.md (hunk 8 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: At the very beginning I make the analogy with HTTP request and response headers, but I can add a bit more detail here if you think it's better.

File: objective-c/auth_sample/README.md (hunk 9 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 7 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 6 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Fixed.

File: objective-c/auth_sample/README.md (hunk 5 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 4 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 3 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 2 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: It is.

File: objective-c/auth_sample/README.md (hunk 1 above)
Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

File: objective-c/auth_sample/README.md (hunk 8 above)
Comment (LisaFC) [inline objective-c/auth_sample/README.md]: Yes, I think that'd be helpful.

Comment (jcanizales) [inline objective-c/auth_sample/README.md]: Done.

**Pull request merged**